app/src/main/java/org/fossasia/pslab/activity/SensorActivity.java

```diff
@@ -140,43 +142,49 @@ public void onItemClick(AdapterView<?> parent, View view, int position, long id)
         });
     }
 
-    private void populateSensors() {
-        ArrayList<Integer> data = new ArrayList<>();
-        dataName.clear();
-        dataAddress.clear();
-        try {
-            tvSensorScan.setText(getResources().getString(R.string.scanning));
-            data = i2c.scan(null);
-        } catch (IOException | NullPointerException e) {
-            e.printStackTrace();
+    private class PopulateSensors extends AsyncTask<Void, Void, Void> {
+        private ArrayList<Integer> data;
+
+        @Override
+        protected Void doInBackground(Void... voids) {
+            data = new ArrayList<>();
+            dataName.clear();
+            dataAddress.clear();
+            try {
+                data = i2c.scan(null);
+            } catch (IOException | NullPointerException e) {
+                e.printStackTrace();
+            }
+            return null;
         }
-        if (data != null) {
-            for (Integer myInt : data) {
-                if (myInt != null && sensorAddr.get(myInt) != null) {
-                    dataAddress.add(String.valueOf(myInt));
+
+        @Override
+        protected void onPostExecute(Void aVoid) {
+            super.onPostExecute(aVoid);
+
+            if (data != null) {
+                for (Integer myInt : data) {
+                    if (myInt != null && sensorAddr.get(myInt) != null) {
+                        dataAddress.add(String.valueOf(myInt));
+                    }
+                }
+                tvData = "";
+
+                for (String s : dataAddress) {
+                    tvData += s + ":" + sensorAddr.get(Integer.parseInt(s)) + "\n";
                 }
-            }
-            tvData = "";
 
-            for (String s : dataAddress) {
-                tvData += s + ":" + sensorAddr.get(Integer.parseInt(s)) + "\n";
+            } else {
+                tvData = getResources().getString(R.string.sensor_not_connected);
             }
 
-        }
-        else {
-            tvData = getResources().getString(R.string.sensor_not_connected);
-        }
+            for (int key : sensorAddr.keySet()) {
+                dataName.add(sensorAddr.get(key));
+            }
 
-        for (int key : sensorAddr.keySet()) {
-            dataName.add(sensorAddr.get(key));
+            tvSensorScan.setText(tvData);
+            adapter.notifyDataSetChanged();
+            buttonSensorAutoScan.setClickable(true);
```

Comment:
Please move setClickable(true) to a place where it will always run. There could be cases where sensorAddr.keySet() could return an empty array and therefore not run setClickable(true). Which then user will be stuck without a button to click

Reply:
It is also not best practice to set clickable in a loop as it takes up iterations doing the same thing again and again, possibly slowing down the program, or in this case, android app

Reply:
@woshikie I  forgot to remove the loop. It is of no use. We can directly use the PopulateSensors class to process thread in the background. So, I have removed the loop and you can now review it.